3p/iframe-transport-client.js

```diff
@@ -33,61 +34,124 @@ export class IframeTransportClient {
     /** @private {!Window} */
     this.win_ = win;
 
-    /** @private {?function(string,string)} */
-    this.listener_ = null;
+    /** @private {Object<string,IframeTransportContext>} */
+    this.creativeIdToContext_ = {};
 
     /** @protected {!IframeMessagingClient} */
-    this.client_ = new IframeMessagingClient(win);
-    this.client_.setHostWindow(this.win_.parent);
-    this.client_.setSentinel(dev().assertString(
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setHostWindow(this.win_.parent);
+    this.iframeMessagingClient_.setSentinel(dev().assertString(
         tryParseJson(this.win_.name)['sentinel'],
         'Invalid/missing sentinel on iframe name attribute' + this.win_.name));
-    this.client_.makeRequest(
+    this.iframeMessagingClient_.makeRequest(
         MessageType.SEND_IFRAME_TRANSPORT_EVENTS,
         MessageType.IFRAME_TRANSPORT_EVENTS,
         eventData => {
           const events =
-              /**
-               * @type
-               * {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
-               */
-              (eventData['events']);
+          /**
+           * @type
+           *   {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
+           */
+          (eventData['events']);
           user().assert(events,
               'Received malformed events list in ' + this.win_.location.href);
           dev().assert(events.length,
               'Received empty events list in ' + this.win_.location.href);
-          user().assert(this.listener_,
-              'Must call onAnalyticsEvent in ' + this.win_.location.href);
           events.forEach(event => {
             try {
-              this.listener_ &&
-                  this.listener_(event.message, event.transportId);
+              user().assert(event.creativeId,
+                  'Received malformed event in ' + this.win_.location.href);
+              const context = this.contextFor_(event.creativeId);
+              context.dispatch(event.message);
             } catch (e) {
               user().error(TAG_,
```

Comment:
Does this really need to be user level?  Why not dev?  Also I believe you can pass the exception as an optional parameter and user/dev logging knows how to handle it such that you see the message (and perhaps some of the stack)

Reply:
Changed to dev(), ditto line 62.